### PR TITLE
Add get_in/1 with safe nil-handling for access and structs

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: lib/elixir/test/elixir/kernel_test.exs [hunk 1]
```diff
@@ -929,6 +929,21 @@ defmodule KernelTest do
...
+      assert get_in(users["john"].age) == 27
+      assert get_in(users["dave"].age) == nil
+      assert get_in(users.meg[:age]) == 23
+      assert get_in(users.meg.age) == 23
+
+      is_nil = nil
+      assert get_in(is_nil.age) == nil
+
+      assert_raise KeyError, ~r"key :unknown not found", fn -> get_in(users.unknown) end
+    end
```
Comment (wojtekmach) [inline lib/elixir/test/elixir/kernel_test.exs:945]: it's probably redundant but I'd consider adding:

```suggestion
     assert_raise KeyError, ~r"key :unknown not found", fn -> get_in(users.meg.unknown) end
    end
```

File: lib/elixir/lib/kernel.ex [hunk 2]
```diff
@@ -2982,6 +2971,63 @@ defmodule Kernel do
...
+  end
+
+  defp nest_get_in(h, _var, []) do
+    h
+  end
+
+  defp nest_get_in(h, var, [{:map, key} | tail]) do
+    quote generated: true do
+      case unquote(h) do
+        %{unquote(key) => unquote(var)} -> unquote(nest_get_in(var, var, tail))
```
Comment (sabiwara) [inline lib/elixir/lib/kernel.ex:3019]: I think we need to bind `key` here?

```elixir
key = john
assert get_in(users[john][:age]) == 27
```

```elixir
     error: undefined variable "john"
     │
 942 │       assert get_in(users[john][:age]) == 27
```

Comment (sabiwara) [inline lib/elixir/lib/kernel.ex:3019]: I think we need to bind `key`?

```elixir
      key = "john"
      assert get_in(users[john][:age]) == 27
```

```elixir
     error: undefined variable "john"
     │
 942 │       assert get_in(users[john][:age]) == 27
 ```

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -2758,13 +2742,18 @@ defmodule Kernel do
 
       get_in(some_struct, [:some_key, :nested_key])
 
-  The good news is that structs have predefined shape. Therefore,
-  you can write instead:
+  There are two alternatives. One is to use regular Elixir syntax
+  for accessing the keys:
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:2746]:
```suggestion
  There are two alternatives. Given structs have predefined keys,
  we can use the `struct.field` notation:
```

File: lib/elixir/lib/kernel.ex [hunk 4]
```diff
@@ -2758,13 +2742,18 @@ defmodule Kernel do
...
 
       some_struct.some_key.nested_key
 
-  If, by any chance, `some_key` can return nil, you can always
-  fallback to pattern matching to provide nested struct handling:
+  However, if you want to safely handle nil values, you can use `get_in/1`:
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex:2750]:
```suggestion
  However, the code above will fail if any of the values return `nil`.
  If you also want to handle nil values, you can use `get_in/1`:
```

**Comment by srcrip** [comment]
@josevalim, would it be feasible to add list handling too if these kinds of changes are on the table? I've been doing something like this a little bit as a utility:

```
  defmodule Utils do
    def get(data, keys)
    def get(data, [h]) when is_function(h), do: h.(:get, data, & &1)
    def get(data, [h | t]) when is_function(h), do: h.(:get, data, &get(&1, t))
    def get(nil, [_]), do: nil
    def get(nil, [_ | t]), do: get(nil, t)

    # list part
    def get(data, [h]) when is_list(data) and is_integer(h), do: Enum.at(data, h)
    def get(data, [h | t]) when is_list(data) and is_integer(h), do: get(Enum.at(data, h), t)

    def get(data, [h]), do: Access.get(data, h)
    def get(data, [h | t]), do: get(Access.get(data, h), t)
  end
```

**Comment by josevalim** [comment]
It has been discussed a couple times. Doing `list[i]` would promote linear access and it is very easy to shoot yourself in the foot. Especially by doing things like `put_in list[i], value` under a loop. I'd try to find the previous discussions.

We have discussed subset operators, inspired by `JSONPath`, such as `get_in(posts[..].comments[..].author.name)` (get the name of all comment authors from all posts) but it comes with other downsides. Definitely a separate convo.

**Comment by josevalim** [comment]
@srcrip that would be a departure of how `*_in` functions work though, since none of them change how the brackets are interpreted and rather just delegate.

Your proposed mechanism would not extensible either (because it would be hardcoded as part of get_in) and the rationale of Access is exactly to make the syntax extensible to everyone. What happens if someone wants to provide a similar for, say, Nx.Tensor?

**Comment by josevalim** [comment]
> Consider the scenario of a user trying to use get_in with a map that contains nested lists. They'll end up decomposing it into a pipeline that calls Enum.at anyways, cause that's what they wanted to do.

Btw, I totally hear you and I agree. However, we need to find a solution that is consistent with the rest of the language. I am not sure at the moment what that is, but let's continue exploring!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Review by gideonfernandes — APPROVED** [review]

**Pull request merged**